pds/contracts/PackNFT_TopShot.cdc

```diff
@@ -0,0 +1,362 @@
+            let v = self._verify(nfts: nfts, salt: salt, commitHash: self.commitHash)
+            self.salt = salt
+            self.status = PackNFT.Status.Revealed
+            emit Revealed(id: id, salt: salt, nfts: v)
+        }
+
+        access(contract) fun open(id: UInt64, nfts: [{IPackNFT.Collectible}]) {
+            assert(self.status == PackNFT.Status.Revealed, message: "Pack status is not Revealed")
+            self._verify(nfts: nfts, salt: self.salt!, commitHash: self.commitHash)
+            self.status = PackNFT.Status.Opened
+            emit Opened(id: id)
+        }
+
+        init(commitHash: String, issuer: Address) {
+            self.commitHash = commitHash
+            self.issuer = issuer
+            self.status = PackNFT.Status.Sealed
+            self.salt = nil
+        }
+    }
+
+    pub resource NFT: NonFungibleToken.INFT, IPackNFT.IPackNFTToken, IPackNFT.IPackNFTOwnerOperator, MetadataViews.Resolver {
+        pub let id: UInt64
+        pub let commitHash: String
+        pub let issuer: Address
+
+        pub fun reveal(openRequest: Bool){
+            PackNFT.revealRequest(id: self.id, openRequest: openRequest)
+        }
+
+        pub fun open(){
+            PackNFT.openRequest(id: self.id)
+        }
+
+       destroy() {
+           let p <- PackNFT.packs.remove(key: self.id) ?? panic("no such pack")
+           PackNFT.totalSupply = PackNFT.totalSupply - (1 as UInt64)
+
+           emit Burned(id: self.id)
+           destroy p
+       }
+
+
+        init(commitHash: String, issuer: Address) {
+            self.id = self.uuid
+            self.commitHash = commitHash
+            self.issuer = issuer
+        }
+
+
+        // All supported metadata views for the Moment including the Core NFT Views
+        //
+        pub fun getViews(): [Type] {
+            return [
+                Type<MetadataViews.Display>(),
+                Type<MetadataViews.ExternalURL>(),
+                Type<MetadataViews.Medias>(),
+                Type<MetadataViews.NFTCollectionData>(),
+                Type<MetadataViews.NFTCollectionDisplay>(),
+                Type<MetadataViews.Royalties>(),
+                Type<MetadataViews.Serial>()
+            ]
+        }
+
+        pub fun resolveView(_ view: Type): AnyStruct? {
+            switch view {
+                case Type<MetadataViews.Display>():
+                    return MetadataViews.Display(
+                        name: "NBA Top Shot Pack",
+                        description: "Reveals official NBA Top Shot Moments when opened",
+                        thumbnail: MetadataViews.HTTPFile(url: self.getImage(imageType: "image", format: "jpeg", width: 256))
+                    )
+                case Type<MetadataViews.ExternalURL>():
+                    return MetadataViews.ExternalURL("https://nbatopshot.com/packnfts/".concat(self.id.toString())) // might have to make a URL that redirects to packs page based on packNFT id -> distribution id
```

Comment:
This url also doesn't exist yet. We will likely need to adjust the FE to enable this to work with the on chain pack id.